crates/trident/src/subsystems/extensions.rs

```diff
@@ -0,0 +1,933 @@
+use std::{
+    fmt::Display,
+    fs, io,
+    path::{Path, PathBuf},
+    time::Duration,
+};
+
+use anyhow::{bail, ensure, Context, Error};
+use log::debug;
+use tempfile::NamedTempFile;
+
+use osutils::{dependencies::Dependency, osrelease::OsRelease};
+use trident_api::{
+    config::Extension,
+    constants::internal_params::COSI_HTTP_CONNECTION_TIMEOUT_SECONDS,
+    error::{InternalError, ReportError, TridentError},
+    primitives::hash::Sha384Hash,
+};
+
+use crate::{
+    engine::{EngineContext, Subsystem},
+    io_utils::{
+        file_reader::FileReader, hashing_reader::HashingReader384, image_streamer::stream_and_hash,
+    },
+};
+
+/// Expected extension-release directory for sysexts
+const SYSEXT_EXTENSION_RELEASE_DIRECTORY: &str = "usr/lib/extension-release.d/";
+/// Expected extension-release directory for confexts
+const CONFEXT_EXTENSION_RELEASE_DIRECTORY: &str = "etc/extension-release.d/";
+
+/// Primary location for storing sysexts on the target OS, relative to the newroot mountpoint
+const DEFAULT_SYSEXT_DIRECTORY: &str = "var/lib/extensions/";
+/// Primary location for storing confexts on the target OS, relative to the newroot mountpoint
+const DEFAULT_CONFEXT_DIRECTORY: &str = "var/lib/confexts/";
+
+/// Temporary directory on target OS for downloading extension images, relative to the newroot mountpoint
+const EXTENSION_IMAGE_DOWNLOAD_DIRECTORY: &str = "var/lib/.extensions-staging/";
```

Comment:
this is not guaranteed to exist in the same filesystem as `/var/lib/extensions/`